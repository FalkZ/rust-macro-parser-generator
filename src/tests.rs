macro_rules! tests {
    (@debug: $name:ident, $t:ident, $str:literal) => {
        #[test]
        fn $name() {
            let r = crate::m1n::renderer::render($str).expect("couldn't compile file");
            insta::assert_debug_snapshot!(&r.$t);
        }
    };
    (@display: $name:ident, $t:ident, $str:literal) => {
        #[test]
        fn $name() {
            let r = crate::m1n::renderer::render($str).expect("couldn't compile file");
            insta::assert_display_snapshot!(&r.$t);
        }
    };
    // `()` indicates that the macro takes no argument.
    ($($name:ident: $str:literal),*) => {
        mod tokens {
            $(
                tests!(@debug: $name, tokens, $str);
            )*
        }
        mod statements {
            $(
                tests!(@debug: $name, statements, $str);
            )*
        }
        mod typescript {
            $(
                tests!(@display: $name, typescript, $str);
            )*
        }
    };
}

tests!(
    enums: "./examples/m1n/Enum.m1n",
    class: "./examples/m1n/Class.m1n",
    file: "./examples/m1n/file.m1n",
    math: "./examples/m1n/std/math.m1n",
    matcher: "./examples/m1n/match.m1n",
    assignment: "./examples/m1n/assignment.m1n",
    operator: "./examples/m1n/operator.m1n"
);
