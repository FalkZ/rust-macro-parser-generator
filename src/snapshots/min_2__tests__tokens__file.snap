---
source: src/tests.rs
expression: "&r.tokens"
---
Tokens {
    index: RefCell {
        value: 139,
    },
    tokens: [
        NEWLINE(
            Token {
                raw: "\n",
                position: Position {
                    column: 0,
                    line: 0,
                },
            },
        ),
        IMPORT(
            Token {
                raw: "import",
                position: Position {
                    column: 0,
                    line: 1,
                },
            },
        ),
        COL(
            Token {
                raw: ":",
                position: Position {
                    column: 6,
                    line: 1,
                },
            },
        ),
        SPACE(
            Token {
                raw: " ",
                position: Position {
                    column: 7,
                    line: 1,
                },
            },
        ),
        NEWLINE(
            Token {
                raw: "\n",
                position: Position {
                    column: 8,
                    line: 1,
                },
            },
        ),
        SPACE(
            Token {
                raw: " ",
                position: Position {
                    column: 0,
                    line: 2,
                },
            },
        ),
        SPACE(
            Token {
                raw: " ",
                position: Position {
                    column: 1,
                    line: 2,
                },
            },
        ),
        SPACE(
            Token {
                raw: " ",
                position: Position {
                    column: 2,
                    line: 2,
                },
            },
        ),
        SPACE(
            Token {
                raw: " ",
                position: Position {
                    column: 3,
                    line: 2,
                },
            },
        ),
        RAWIDENT(
            Token {
                raw: "`@std/util`",
                position: Position {
                    column: 4,
                    line: 2,
                },
            },
        ),
        CBRACKETOPEN(
            Token {
                raw: "{",
                position: Position {
                    column: 15,
                    line: 2,
                },
            },
        ),
        SPACE(
            Token {
                raw: " ",
                position: Position {
                    column: 16,
                    line: 2,
                },
            },
        ),
        IDENT(
            Token {
                raw: "pipe",
                position: Position {
                    column: 17,
                    line: 2,
                },
            },
        ),
        SPACE(
            Token {
                raw: " ",
                position: Position {
                    column: 21,
                    line: 2,
                },
            },
        ),
        CBRACKETCLOSE(
            Token {
                raw: "}",
                position: Position {
                    column: 22,
                    line: 2,
                },
            },
        ),
        COMMA(
            Token {
                raw: ",",
                position: Position {
                    column: 23,
                    line: 2,
                },
            },
        ),
        NEWLINE(
            Token {
                raw: "\n",
                position: Position {
                    column: 24,
                    line: 2,
                },
            },
        ),
        SPACE(
            Token {
                raw: " ",
                position: Position {
                    column: 0,
                    line: 3,
                },
            },
        ),
        SPACE(
            Token {
                raw: " ",
                position: Position {
                    column: 1,
                    line: 3,
                },
            },
        ),
        SPACE(
            Token {
                raw: " ",
                position: Position {
                    column: 2,
                    line: 3,
                },
            },
        ),
        SPACE(
            Token {
                raw: " ",
                position: Position {
                    column: 3,
                    line: 3,
                },
            },
        ),
        RAWIDENT(
            Token {
                raw: "`@std/math`",
                position: Position {
                    column: 4,
                    line: 3,
                },
            },
        ),
        SEMI(
            Token {
                raw: ";",
                position: Position {
                    column: 15,
                    line: 3,
                },
            },
        ),
        NEWLINE(
            Token {
                raw: "\n",
                position: Position {
                    column: 16,
                    line: 3,
                },
            },
        ),
        IDENT(
            Token {
                raw: "version",
                position: Position {
                    column: 0,
                    line: 4,
                },
            },
        ),
        COL(
            Token {
                raw: ":",
                position: Position {
                    column: 7,
                    line: 4,
                },
            },
        ),
        SPACE(
            Token {
                raw: " ",
                position: Position {
                    column: 8,
                    line: 4,
                },
            },
        ),
        NUMBER(
            Token {
                raw: "0",
                position: Position {
                    column: 9,
                    line: 4,
                },
            },
        ),
        SEMI(
            Token {
                raw: ";",
                position: Position {
                    column: 10,
                    line: 4,
                },
            },
        ),
        NEWLINE(
            Token {
                raw: "\n",
                position: Position {
                    column: 11,
                    line: 4,
                },
            },
        ),
        IDENT(
            Token {
                raw: "derive",
                position: Position {
                    column: 0,
                    line: 5,
                },
            },
        ),
        COL(
            Token {
                raw: ":",
                position: Position {
                    column: 6,
                    line: 5,
                },
            },
        ),
        SPACE(
            Token {
                raw: " ",
                position: Position {
                    column: 7,
                    line: 5,
                },
            },
        ),
        IDENT(
            Token {
                raw: "equal",
                position: Position {
                    column: 8,
                    line: 5,
                },
            },
        ),
        SEMI(
            Token {
                raw: ";",
                position: Position {
                    column: 13,
                    line: 5,
                },
            },
        ),
        NEWLINE(
            Token {
                raw: "\n",
                position: Position {
                    column: 14,
                    line: 5,
                },
            },
        ),
        NEWLINE(
            Token {
                raw: "\n",
                position: Position {
                    column: 0,
                    line: 6,
                },
            },
        ),
        IDENT(
            Token {
                raw: "PI",
                position: Position {
                    column: 0,
                    line: 7,
                },
            },
        ),
        SPACE(
            Token {
                raw: " ",
                position: Position {
                    column: 2,
                    line: 7,
                },
            },
        ),
        EQUAL(
            Token {
                raw: "=",
                position: Position {
                    column: 3,
                    line: 7,
                },
            },
        ),
        SPACE(
            Token {
                raw: " ",
                position: Position {
                    column: 4,
                    line: 7,
                },
            },
        ),
        IDENT(
            Token {
                raw: "a",
                position: Position {
                    column: 5,
                    line: 7,
                },
            },
        ),
        SPACE(
            Token {
                raw: " ",
                position: Position {
                    column: 6,
                    line: 7,
                },
            },
        ),
        NEWLINE(
            Token {
                raw: "\n",
                position: Position {
                    column: 7,
                    line: 7,
                },
            },
        ),
        SPACE(
            Token {
                raw: " ",
                position: Position {
                    column: 0,
                    line: 8,
                },
            },
        ),
        SPACE(
            Token {
                raw: " ",
                position: Position {
                    column: 1,
                    line: 8,
                },
            },
        ),
        SPACE(
            Token {
                raw: " ",
                position: Position {
                    column: 2,
                    line: 8,
                },
            },
        ),
        SPACE(
            Token {
                raw: " ",
                position: Position {
                    column: 3,
                    line: 8,
                },
            },
        ),
        PLUS(
            Token {
                raw: "+",
                position: Position {
                    column: 4,
                    line: 8,
                },
            },
        ),
        SPACE(
            Token {
                raw: " ",
                position: Position {
                    column: 5,
                    line: 8,
                },
            },
        ),
        IDENT(
            Token {
                raw: "b",
                position: Position {
                    column: 6,
                    line: 8,
                },
            },
        ),
        NEWLINE(
            Token {
                raw: "\n",
                position: Position {
                    column: 7,
                    line: 8,
                },
            },
        ),
        SPACE(
            Token {
                raw: " ",
                position: Position {
                    column: 0,
                    line: 9,
                },
            },
        ),
        SPACE(
            Token {
                raw: " ",
                position: Position {
                    column: 1,
                    line: 9,
                },
            },
        ),
        SPACE(
            Token {
                raw: " ",
                position: Position {
                    column: 2,
                    line: 9,
                },
            },
        ),
        SPACE(
            Token {
                raw: " ",
                position: Position {
                    column: 3,
                    line: 9,
                },
            },
        ),
        PLUS(
            Token {
                raw: "+",
                position: Position {
                    column: 4,
                    line: 9,
                },
            },
        ),
        SPACE(
            Token {
                raw: " ",
                position: Position {
                    column: 5,
                    line: 9,
                },
            },
        ),
        NUMBER(
            Token {
                raw: "0",
                position: Position {
                    column: 6,
                    line: 9,
                },
            },
        ),
        DOT(
            Token {
                raw: ".",
                position: Position {
                    column: 7,
                    line: 9,
                },
            },
        ),
        NUMBER(
            Token {
                raw: "012",
                position: Position {
                    column: 8,
                    line: 9,
                },
            },
        ),
        SEMI(
            Token {
                raw: ";",
                position: Position {
                    column: 11,
                    line: 9,
                },
            },
        ),
        NEWLINE(
            Token {
                raw: "\n",
                position: Position {
                    column: 12,
                    line: 9,
                },
            },
        ),
        NEWLINE(
            Token {
                raw: "\n",
                position: Position {
                    column: 0,
                    line: 10,
                },
            },
        ),
        PUB(
            Token {
                raw: "pub",
                position: Position {
                    column: 0,
                    line: 11,
                },
            },
        ),
        SPACE(
            Token {
                raw: " ",
                position: Position {
                    column: 3,
                    line: 11,
                },
            },
        ),
        MUT(
            Token {
                raw: "mut",
                position: Position {
                    column: 4,
                    line: 11,
                },
            },
        ),
        SPACE(
            Token {
                raw: " ",
                position: Position {
                    column: 7,
                    line: 11,
                },
            },
        ),
        IDENT(
            Token {
                raw: "Pi",
                position: Position {
                    column: 8,
                    line: 11,
                },
            },
        ),
        SPACE(
            Token {
                raw: " ",
                position: Position {
                    column: 10,
                    line: 11,
                },
            },
        ),
        EQUAL(
            Token {
                raw: "=",
                position: Position {
                    column: 11,
                    line: 11,
                },
            },
        ),
        SPACE(
            Token {
                raw: " ",
                position: Position {
                    column: 12,
                    line: 11,
                },
            },
        ),
        NUMBER(
            Token {
                raw: "123",
                position: Position {
                    column: 13,
                    line: 11,
                },
            },
        ),
        SEMI(
            Token {
                raw: ";",
                position: Position {
                    column: 16,
                    line: 11,
                },
            },
        ),
        NEWLINE(
            Token {
                raw: "\n",
                position: Position {
                    column: 17,
                    line: 11,
                },
            },
        ),
        NEWLINE(
            Token {
                raw: "\n",
                position: Position {
                    column: 0,
                    line: 12,
                },
            },
        ),
        IDENT(
            Token {
                raw: "fn",
                position: Position {
                    column: 0,
                    line: 13,
                },
            },
        ),
        BRACKETOPEN(
            Token {
                raw: "(",
                position: Position {
                    column: 2,
                    line: 13,
                },
            },
        ),
        IDENT(
            Token {
                raw: "a",
                position: Position {
                    column: 3,
                    line: 13,
                },
            },
        ),
        COMMA(
            Token {
                raw: ",",
                position: Position {
                    column: 4,
                    line: 13,
                },
            },
        ),
        SPACE(
            Token {
                raw: " ",
                position: Position {
                    column: 5,
                    line: 13,
                },
            },
        ),
        IDENT(
            Token {
                raw: "b",
                position: Position {
                    column: 6,
                    line: 13,
                },
            },
        ),
        BRACKETCLOSE(
            Token {
                raw: ")",
                position: Position {
                    column: 7,
                    line: 13,
                },
            },
        ),
        SPACE(
            Token {
                raw: " ",
                position: Position {
                    column: 8,
                    line: 13,
                },
            },
        ),
        EQUAL(
            Token {
                raw: "=",
                position: Position {
                    column: 9,
                    line: 13,
                },
            },
        ),
        SPACE(
            Token {
                raw: " ",
                position: Position {
                    column: 10,
                    line: 13,
                },
            },
        ),
        IDENT(
            Token {
                raw: "a",
                position: Position {
                    column: 11,
                    line: 13,
                },
            },
        ),
        SPACE(
            Token {
                raw: " ",
                position: Position {
                    column: 12,
                    line: 13,
                },
            },
        ),
        NEWLINE(
            Token {
                raw: "\n",
                position: Position {
                    column: 13,
                    line: 13,
                },
            },
        ),
        SPACE(
            Token {
                raw: " ",
                position: Position {
                    column: 0,
                    line: 14,
                },
            },
        ),
        SPACE(
            Token {
                raw: " ",
                position: Position {
                    column: 1,
                    line: 14,
                },
            },
        ),
        SPACE(
            Token {
                raw: " ",
                position: Position {
                    column: 2,
                    line: 14,
                },
            },
        ),
        SPACE(
            Token {
                raw: " ",
                position: Position {
                    column: 3,
                    line: 14,
                },
            },
        ),
        PLUS(
            Token {
                raw: "+",
                position: Position {
                    column: 4,
                    line: 14,
                },
            },
        ),
        SPACE(
            Token {
                raw: " ",
                position: Position {
                    column: 5,
                    line: 14,
                },
            },
        ),
        IDENT(
            Token {
                raw: "b",
                position: Position {
                    column: 6,
                    line: 14,
                },
            },
        ),
        SEMI(
            Token {
                raw: ";",
                position: Position {
                    column: 7,
                    line: 14,
                },
            },
        ),
        NEWLINE(
            Token {
                raw: "\n",
                position: Position {
                    column: 8,
                    line: 14,
                },
            },
        ),
        NEWLINE(
            Token {
                raw: "\n",
                position: Position {
                    column: 0,
                    line: 15,
                },
            },
        ),
        IDENT(
            Token {
                raw: "one",
                position: Position {
                    column: 0,
                    line: 16,
                },
            },
        ),
        BRACKETOPEN(
            Token {
                raw: "(",
                position: Position {
                    column: 3,
                    line: 16,
                },
            },
        ),
        IDENT(
            Token {
                raw: "b",
                position: Position {
                    column: 4,
                    line: 16,
                },
            },
        ),
        BRACKETCLOSE(
            Token {
                raw: ")",
                position: Position {
                    column: 5,
                    line: 16,
                },
            },
        ),
        SPACE(
            Token {
                raw: " ",
                position: Position {
                    column: 6,
                    line: 16,
                },
            },
        ),
        EQUAL(
            Token {
                raw: "=",
                position: Position {
                    column: 7,
                    line: 16,
                },
            },
        ),
        SPACE(
            Token {
                raw: " ",
                position: Position {
                    column: 8,
                    line: 16,
                },
            },
        ),
        IDENT(
            Token {
                raw: "console",
                position: Position {
                    column: 9,
                    line: 16,
                },
            },
        ),
        DOT(
            Token {
                raw: ".",
                position: Position {
                    column: 16,
                    line: 16,
                },
            },
        ),
        IDENT(
            Token {
                raw: "log",
                position: Position {
                    column: 17,
                    line: 16,
                },
            },
        ),
        BRACKETOPEN(
            Token {
                raw: "(",
                position: Position {
                    column: 20,
                    line: 16,
                },
            },
        ),
        IDENT(
            Token {
                raw: "b",
                position: Position {
                    column: 21,
                    line: 16,
                },
            },
        ),
        BRACKETCLOSE(
            Token {
                raw: ")",
                position: Position {
                    column: 22,
                    line: 16,
                },
            },
        ),
        SEMI(
            Token {
                raw: ";",
                position: Position {
                    column: 23,
                    line: 16,
                },
            },
        ),
        NEWLINE(
            Token {
                raw: "\n",
                position: Position {
                    column: 24,
                    line: 16,
                },
            },
        ),
        NEWLINE(
            Token {
                raw: "\n",
                position: Position {
                    column: 0,
                    line: 17,
                },
            },
        ),
        PUB(
            Token {
                raw: "pub",
                position: Position {
                    column: 0,
                    line: 18,
                },
            },
        ),
        SPACE(
            Token {
                raw: " ",
                position: Position {
                    column: 3,
                    line: 18,
                },
            },
        ),
        IDENT(
            Token {
                raw: "zero",
                position: Position {
                    column: 4,
                    line: 18,
                },
            },
        ),
        BRACKETOPEN(
            Token {
                raw: "(",
                position: Position {
                    column: 8,
                    line: 18,
                },
            },
        ),
        BRACKETCLOSE(
            Token {
                raw: ")",
                position: Position {
                    column: 9,
                    line: 18,
                },
            },
        ),
        SPACE(
            Token {
                raw: " ",
                position: Position {
                    column: 10,
                    line: 18,
                },
            },
        ),
        EQUAL(
            Token {
                raw: "=",
                position: Position {
                    column: 11,
                    line: 18,
                },
            },
        ),
        SPACE(
            Token {
                raw: " ",
                position: Position {
                    column: 12,
                    line: 18,
                },
            },
        ),
        NEWLINE(
            Token {
                raw: "\n",
                position: Position {
                    column: 13,
                    line: 18,
                },
            },
        ),
        SPACE(
            Token {
                raw: " ",
                position: Position {
                    column: 0,
                    line: 19,
                },
            },
        ),
        SPACE(
            Token {
                raw: " ",
                position: Position {
                    column: 1,
                    line: 19,
                },
            },
        ),
        SPACE(
            Token {
                raw: " ",
                position: Position {
                    column: 2,
                    line: 19,
                },
            },
        ),
        SPACE(
            Token {
                raw: " ",
                position: Position {
                    column: 3,
                    line: 19,
                },
            },
        ),
        IDENT(
            Token {
                raw: "a",
                position: Position {
                    column: 4,
                    line: 19,
                },
            },
        ),
        SPACE(
            Token {
                raw: " ",
                position: Position {
                    column: 5,
                    line: 19,
                },
            },
        ),
        NEWLINE(
            Token {
                raw: "\n",
                position: Position {
                    column: 6,
                    line: 19,
                },
            },
        ),
        SPACE(
            Token {
                raw: " ",
                position: Position {
                    column: 0,
                    line: 20,
                },
            },
        ),
        SPACE(
            Token {
                raw: " ",
                position: Position {
                    column: 1,
                    line: 20,
                },
            },
        ),
        SPACE(
            Token {
                raw: " ",
                position: Position {
                    column: 2,
                    line: 20,
                },
            },
        ),
        SPACE(
            Token {
                raw: " ",
                position: Position {
                    column: 3,
                    line: 20,
                },
            },
        ),
        PLUS(
            Token {
                raw: "+",
                position: Position {
                    column: 4,
                    line: 20,
                },
            },
        ),
        SPACE(
            Token {
                raw: " ",
                position: Position {
                    column: 5,
                    line: 20,
                },
            },
        ),
        IDENT(
            Token {
                raw: "b",
                position: Position {
                    column: 6,
                    line: 20,
                },
            },
        ),
        SEMI(
            Token {
                raw: ";",
                position: Position {
                    column: 7,
                    line: 20,
                },
            },
        ),
    ],
}
